pilot/pkg/simulation/traffic.go

```diff
@@ -514,10 +516,17 @@ func (sim *Simulation) matchFilterChain(chains []*listener.FilterChain, defaultC
 	}, func(fc *listener.FilterChainMatch) bool {
 		return sets.NewSet(fc.GetApplicationProtocols()...).Contains(input.Alpn)
 	})
+
```

Comment:
Maybe it would be more clear to add the overlapping check there, then modify the filter() function to return the best match instead of all matches? I think that would mirror envoy logic better. 

We may actually have false matches today. For example, if I have 

```
- match: sni: *.com, transport: tls
- match: sni: example.com, transport: plain
```

and I send a request that is `sni: example.com, transport: tls`. Current logic would match FC 0; envoy would amtch neither *I think*

Reply:
> Your comment mentioned checking for overlapping FC when we create the objects instead of during match, but I don't see anything do that. is that intended? I think we could improve validateFilterChainMatch

No, it is not included in this patchset. I did think that it should go that way but I was intended not to break too much on the current logic for this first attempt.
I agree that we could improve validateFilterChainMatch. What's in my mind is roughly similar to what you've suggested below. :)

Reply:
> Maybe it would be more clear to add the overlapping check there, then modify the filter() function to return the best match instead of all matches? I think that would mirror envoy logic better.
> 
> We may actually have false matches today. For example, if I have
> 
> ```
> - match: sni: *.com, transport: tls
> - match: sni: example.com, transport: plain
> ```
> 
> and I send a request that is `sni: example.com, transport: tls`. Current logic would match FC 0; envoy would amtch neither _I think_

Thanks a lot for the hints! I'll give a try for this approach.